torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -749,4 +749,79 @@ xla::XlaOp BuildMaskedScatter(xla::XlaOp input, xla::XlaOp mask,
       scatter_dnums);
 }
 
+std::vector<xla::XlaOp> BuildAmpForachNonFiniteCheckAndUnscale(
+    const std::vector<xla::XlaOp>& inputs, const xla::XlaOp& found_inf_float,
+    const xla::XlaOp& inv_scale) {
+  const xla::PrimitiveType origin_type =
+      XlaHelpers::ShapeOfXlaOp(found_inf_float).element_type();
+  std::vector<xla::XlaOp> found_infs;
+  xla::XlaOp one = xla::One(inputs[0].builder(), xla::PrimitiveType::S32);
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    xla::XlaOp all_finite =
+        xla::ReduceAll(xla::ConvertElementType(xla::IsFinite(inputs[i]),
+                                               xla::PrimitiveType::S32),
+                       one,
+                       xla::CreateScalarAndComputation(xla::PrimitiveType::S32,
+                                                       inputs[i].builder()));
+    found_infs.push_back(one - all_finite);
+  }
+  xla::XlaOp found_inf =
+      xla::ConvertElementType(found_inf_float, xla::PrimitiveType::S32);
+  for (size_t i = 0; i < found_infs.size(); ++i) {
+    found_inf = xla::Or(found_inf, found_infs[i]);
+  }
+  std::vector<xla::XlaOp> results;
+
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    results.push_back(inputs[i] * inv_scale);
+  }
+  results.push_back(xla::ConvertElementType(found_inf, origin_type));
+  return results;
+}
+
+std::vector<xla::XlaOp> BuildAmpUpdateScale(const xla::XlaOp& growth_tracker,
+                                            const xla::XlaOp& current_scale,
+                                            const xla::XlaOp& found_inf_float,
+                                            float scale_growth_factor,
+                                            float scale_backoff_factor,
+                                            int scale_growth_interval) {
+  xla::XlaOp one = xla::One(growth_tracker.builder(), xla::PrimitiveType::S32);
+  xla::XlaOp one_float =
+      xla::One(growth_tracker.builder(), xla::PrimitiveType::F32);
+  xla::XlaOp found_inf = xla::Min(
+      xla::ConvertElementType(found_inf_float, xla::PrimitiveType::S32), one);
```

Comment:
I will leave it to you to fix the rest of the code, but I think it is better to leave the boolean value as `xla::PrimitiveType::PRED` and use `AND` `OR`

Reply:
Done.